File: bukkit/libs/misc/src/main/java/dialight/observable/ObservableObject.java
package dialight.observable;

import java.util.*;
import java.util.function.BiConsumer;

public class ObservableObject<V> {

    protected V value;
    private final Map<Object, BiConsumer<V, V>> onChange = new LinkedHashMap<>();

    public ObservableObject() {
        this.value = null;
    }
    public ObservableObject(V value) {
        this.value = value;
    }

    protected void fireChange(V fr, V to) {
        for(BiConsumer<V, V> op : onChange.values()) op.accept(fr, to);
    }

    public ObservableObject<V> onChange(Object key, BiConsumer<V, V> op) {
        onChange.put(key, op);
        return this;
    }
    public ObservableObject<V> removeListeners(Object key) {
        onChange.remove(key);
        return this;
    }

    public V getValue() {
        return value;
    }

    public void setValue(V value) {
        V oldValue = this.value;
        if(Objects.equals(value, oldValue)) return;
        this.value = value;
        fireChange(oldValue, value);
    }

    public ObservableObject<V> asImmutable() {
        return new ImmutableObservableObject<>(this);
    }

    @Override public String toString() {
        return this.value.toString();
    }

}

File: sponge/libs/misc/src/main/kotlin/dialight/observable/ObservableObject.kt
package dialight.observable

import java.util.*

class ObservableObject<V>(value: V) {

    private val onChange: MutableCollection<(V, V) -> Unit> = LinkedList()

    protected fun fireChange(fr: V, to: V) {
        for(op in onChange) op(fr, to)
    }

    fun onChange(op: (V, V) -> Unit) : ObservableObject<V> {
        onChange += op
        return this
    }

    var value: V = value
        set(value) {
            val oldValue = field
            field = value
            fireChange(oldValue, value)
        }

}